### ISSUE TITLE
Data Generation with GPT-2 

### ISSUE THREAD

**Issue by Husamx** [RasaHQ/rasa#9822]
### What problem are you trying to solve?

I was a user of Rasa back in 2018 and early 2019\. One of the major issues that I faced when building a chatbot was the lack of data that I had for training the intent classifier. We do know that collecting as much real-world data as possible would of course be the optimal solution here, but this may not be possible for everyone. We can agree that one way to collect data would be to start with a suboptimal classifier, trained on any available training data, e.g. manually scripted data, and then collect real-world data as users engage with the final chatbot. Of course there are multiple issues with this process, such as privacy concerns and limits to the available resources, e.g. time and labor, needed to create the initial training data. For this reason, I've been researching ways to speed up the data creation process, so that it becomes easier for the user to train and improve the initial classifier. In the suggested approach, we take advantage of models like GPT-2 to generate synthetic training data.

### What's your suggested solution?

Last year, I published the paper "Textual Data Augmentation for Efficient Active Learning on Tiny Datasets" at EMNLP 2020\. In this work, we proposed a method in which a classifier (e.g. intent classifier) and a language generation model e.g. GPT-2, can be finetuned on the existing dataset (this can be a small sample of handcrafted data). GPT-2 is then used to generate synthetic data examples. Once an example is generated, we would need to know its usefulness to the classifier. Intuitively, if the classifier already predicts a sample with high confidence, then there is little to gain if we retrain the same classifier with this example. However, hard examples for which the model has less confidence would be of interest here. If these samples are added to the training data, then we would expect an improved performance. For this, we compute the classifier's entropy for each generated example and focus on the ones with high entropy scores. We will still need to make sure that the generated examples are well-formed and correctly labeled. Since we cannot rely on the classifier's pseudo labels, we manually review a sample of the generated data (e.g. top 50 examples by entropy). The classifier can then be retrained on the newly generated data (initial training data + reviewed examples). This process can then be repeated over multiple active learning runs until a convergence criterion is satisfied. In the paper we implemented 2 approaches for the data generation process, one is based on top-k random sampling, named it Non-Guided Data Generation NGDG), and the other is with the use of Monte Carlo Tree Search (MCTS). We tested this approach on SST-2 and TREC-6 datasets and managed to achieve impressive improvements. It would be interesting to see how well this approach would work for Rasa users in real life settings.

### Examples (if relevant)

 _No response_ 

### Is anything blocking this from being implemented? (if relevant)

 _No response_ 

### Definition of Done

 _No response_
